### DISCUSSION TITLE
Is Cosmos database supported by this package?

### DISCUSSION THREAD

**Discussion by azthrun** [Azure/azure-mobile-apps#308] [Q&A] [answered]
Hello,

Most of my applications are using Cosmos as the backend database.
I wonder if this mobile sync package is, or will be, supporting Cosmos.

If Cosmos is supported, do I have an option to use Cosmos SDK instead of EFCore Cosmos provider?
Because of the current limitations on the EFCore Cosmos provider, my APIs are using the Cosmos SDK at the moment.

Any suggestions on how to implement an API using Azure mobile app with Cosmos are appreciated.

**Comment by adrianhall** [comment, marked as answer]
It should be possible.  You have to create an implementation of the `IRepository<T>` for Cosmos SDK.  The SDK has EF Core and In memory providers you can use as an example.  (If you do create a generic one, please let me know so we can advertise it for you!)  As you will see, the `IRepository<T>` is just a CRUD plus an IQueryable for LINQ usage (which may be the hardest piece of the repository pattern).

Underneath, the OData v4 translation uses IQueryable to produce query sets; it uses `repository.AsQueryable().{some set of LINQ extension methods}.ToList()` to do the slicing, so it's important that your LINQ provider implements the subset of LINQ queries.  Although I have not tried it, [`GetItemLinqQueryable`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.container.getitemlinqqueryable?view=azure-dotnet) is probably what you want.

**Reply by AathifMahir** [reply]
Does offline sync also support CosmosDB? @adrianhall

**Reply by adrianhall** [reply]
Offline Sync doesn't care what database you have behind the scenes.

The problem you will run into is that the offline store is SQLite (table driven with defined fields) and CosmosDB is NoSQL (no pre-defined schema).  You have to define a schema for the translation required for offline sync.

**Reply by adrianhall** [reply]
Also, note that I've included information in the documentation and a sample in the samples area on how to do offline sync with CosmosDB.

**Reply by AathifMahir** [reply]
> Also, note that I've included information in the documentation and a sample in the samples area on how to do offline sync with CosmosDB.

Would you mind sharing the url of documentation? Can't seem to find anything related to CosmosDB in the documentation

**Reply by AathifMahir** [reply]
@adrianhall Also, is there a possibility to cache Azure Cosmos DB without DataSync in favor of using azure.cosmoss SDK instead EF Core. Since I was using this on client application mostly I don't need the features on EF Core, Using Azure Cosmos SDK is more than Enough in that case how we can achieve the offline data persistence

**Reply by adrianhall** [reply]
That's a longer question :-)  Short version, yes, but it's hard.

You can write an IRepository implementation for Cosmos that uses the CosmosDB SDK.  However, the list operation requires the dataset to be presented as an async LINQ endpoint that can be modified prior to execution (so that the query is run on the backend and not filtered within the ASP.NET Core app).  The list operation is used to handle incremental sync (only get the items since the last update), paging, and client-based filtering (only get the items that my client needs).  

Good news is that [Azure.Cosmos includes LINQ](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/linq-to-sql).  However, I have not validated the library as being suitable for all cases with Azure Mobile Apps.

**Reply by AathifMahir** [reply]
> That's a longer question :-)  Short version, yes, but it's hard.
> 
> You can write an IRepository implementation for Cosmos that uses the CosmosDB SDK.  However, the list operation requires the dataset to be presented as an async LINQ endpoint that can be modified prior to execution (so that the query is run on the backend and not filtered within the ASP.NET Core app).  The list operation is used to handle incremental sync (only get the items since the last update), paging, and client-based filtering (only get the items that my client needs).  
> 
> Good news is that [Azure.Cosmos includes LINQ](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/linq-to-sql).  However, I have not validated the library as being suitable for all cases with Azure Mobile Apps.

Thanks, as far per my requirements I guess it's best to develop a class library that targets .net Standard and publish that in Nuget, therefore it could help others as well.

As per the Specs of the implementation, I will share that once I decided on this is the way forward, would be helpful if you can provide a feedback on Specs. Thanks again

**Reply by AathifMahir** [reply]
> Also, note that I've included information in the documentation and a sample in the samples area on how to do offline sync with CosmosDB.

By the way, would you mind sharing url for documentation for Azure Mobile Apps Offline Sync for CosmosDB
